### ISSUE TITLE
Weird behaviour using DependencyMatcher with OP attribute

### ISSUE THREAD

**Issue by werew** [explosion/spaCy#6661]
## How to reproduce the behaviour

From the doc it is not very clear how the OP attribute should behave when used in the DependencyMatcher or if it is not even supported.

Take this example:

```
import spacy
from spacy.matcher import DependencyMatcher


nlp = spacy.load("en_core_web_sm")

text = "The dress is beautiful"
doc = nlp(text)

matcher = DependencyMatcher(nlp.vocab)
pattern = [
    {"RIGHT_ID": "is", "RIGHT_ATTRS": {"LEMMA": "be"}},
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "subj",
        "RIGHT_ATTRS": {"DEP": "nsubj"},
    },
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "adj",
        "RIGHT_ATTRS": {"POS": "ADJ",},
    },
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "not",
        "RIGHT_ATTRS": {"DEP": "neg", "OP": "?"},
    },
]
matcher.add("test", [pattern])


for _, tokens in matcher(doc):
    print("------------")
    print(tokens)
    for token in tokens:
        print(doc[token])
```

This will print out:

```
------------
[2, 1, 3, 1]
is
dress
beautiful
dress
------------
[2, 1, 3, 3]
is
dress
beautiful
beautiful
```

The last token in the pattern (called "not") is never matched, however we still get two matches where the 4th entry is an unrelated token. Using the `*` operators gives the same result.
Matching the sentence "The dress is not beautiful" still gives a weird result:

```
------------
[2, 1, 4, 1]
is
dress
beautiful
dress
------------
[2, 1, 4, 3]
is
dress
beautiful
not
------------
[2, 1, 4, 4]
is
dress
beautiful
beautiful
```

## Info about spaCy

- **spaCy version:** 3.0.0rc2
- **Platform:** Linux-5.10.3-arch1-1-x86_64-with-glibc2.2.5
- **Python version:** 3.8.7
- **Pipelines:** en_core_web_sm (3.0.0a0), en_core_web_trf (3.0.0a0)

**Comment by werew** [comment]
I also tried setting the OP attribute to "!" in order to match all trees not containing a negation.
```
pattern = [
    {"RIGHT_ID": "is", "RIGHT_ATTRS": {"LEMMA": "be"}},
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "subj",
        "RIGHT_ATTRS": {"DEP": "nsubj"},
    },
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "adj",
        "RIGHT_ATTRS": {"POS": "ADJ",},
    },
    {
        "LEFT_ID": "is",
        "REL_OP": ">",
        "RIGHT_ID": "not",
        "RIGHT_ATTRS": {"DEP": "neg", "OP": "!"},
    },
]
```
When matching the sentence "The dress is not beautiful" this still gives two matches:

```
------------
[2, 1, 4, 1]
is
dress
beautiful
dress
------------
[2, 1, 4, 4]
is
dress
beautiful
beautiful
```
I believe there should be no matches. For reference here is how the dependency tree looks like:

![image](https://user-images.githubusercontent.com/10875013/103482466-f4301a80-4dd8-11eb-8663-10e70f009868.png)

**Comment by adrianeboyd** [comment]
I think the current dependency matcher algorithm is only going to work correctly if the token pattern matches exactly one token. I think, at least initially, we should simply forbid `OP` in the patterns so you don't get these kinds of weird results.

**Comment by werew** [comment]
> I think, at least initially, we should simply forbid OP in the patterns so you don't get these kinds of weird results.
Sounds reasonable :+1: 

The "!" OP value could be a nice to have, other OP values can be replaced by using multiple patterns, but not "!" (though the callback can handle that).

For the future it would be really cool to have DependencyMatcher return a list of lists of token ids instead of a simple list of token ids. In this case each sub-list would contain all token ids matched by the token pattern at the corresponding position (so the sublist can be potentially empty if OP is in [!,?,*], or contain multiple elements for OP is in [+,*,]).

**Comment by adrianeboyd** [comment]
I think part of this is that `"OP": "?"` is returning buggy matches in some cases: #6677. We would want to support `"OP": "!"`, so I think we'd need to add more restricted pattern validation to prohibit the other operators and add some checks that the matches are all exactly one token long.

I don't think `+` or `*` make sense in the token pattern part of these patterns and returning lists of tokens wouldn't make sense for most kinds of pattern structures. (Lists would only maybe make sense for the leaf nodes in the pattern, I think?)

It would be possible to extend the dependency pattern schema to allow additional operators at the relation level, but I don't think it's a good idea at the token level.

**Comment by werew** [comment]
> I don't think + or * make sense in the token pattern part of these patterns and returning lists of tokens wouldn't make sense for most kinds of pattern structures. (Lists would only maybe make sense for the leaf nodes in the pattern, I think?)

I agree that most patterns will not need it. There are however instances in which they might become useful:

![image](https://user-images.githubusercontent.com/10875013/103836060-d0423280-507f-11eb-8037-92b8db026f8d.png)

At the end it is a design choice. Considering the low impact I think it is wise to keep things simple for now, users can still do any custom handling in the callback to handle those special cases.

**Comment by adrianeboyd** [comment]
Closed by #6824. (The linked issue auto-closing has been flaky for a while.)
